design-proposals/2022-11-29-metal-port.md

```diff
@@ -0,0 +1,447 @@
+# MapLibre Metal Port Design Proposal
+
+Before we dive in, a bit about our process.  Stamen Design, with a sub-contract to Wet Dog Weather, has been contracted by Amazon (AWS) to develop two proposals to upgrade MapLibre Native.  The first was the Rendering Modularization Proposal.  This one is the Metal Port.
+
+The mechanism for this will be a Pull Request, which we have obviously opened.  This will result in as much discussion as the community would like here, on the OSM Slack and by email or video call (we're available, so reach out).  As we wrap up our specific proposal, that discussion will hopefully reach a consensus and we'll be ready for a Yes or No on the PR by the end of the year.  
+
+## Motivation
+
+MapLibre Native is currently depending on a deprecated rendering SDK (OpenGL) for iOS. It needs to move to Metal for iOS in some form.  Thus the graphics implementations must diverge, either by doing so within MapLibre Native or depending on another toolkit to do the rendering entirely. 
+
+We are proposing the former approach, with support for multiple rendering SDKs in the MapLibre Native toolkit itself. That is discussed in more detail in the MapLibre Rendering Modularization Design Plan.  This proposal assumes that one is adopted.
+
+But before that, a quick note on terminology.
+- "toolkit" usually refers to MapLibre Native, but may also be used in context for another project at a similar level.
+- "SDK" or Software Development Kit usually refers to the rendering library, like OpenGL or Metal, at least in this document.
+- "API" refers to an Application Programming Interface.  In this context it refers to the classes and methods made available to a developer.
+
+## Goals
+
+It is useful to split our goals into three sections to articulate what this proposal aims to accomplish:
+
+- **North Stars:** These guide the overall direction of the proposal, and are the overarching _why_.
+
+- **Core Functionality:** These are concrete changes we are proposing to the library to get to our North Stars.
+
+- **Evaluation metrics:** These are specific criteria we will accomplish through this proposal.
+
+### <a name="north-stars">North Stars</a>
+1. The MapLibre Native toolkit should use Metal for rendering on iOS
+2. The MapLibre Native toolkit should use Metal for rendering on MacOS
+
+### <a name="core">Core Functionality</a>
+1. The OpenGL renderer will still be available on iOS as a compile time flag
+2. Shaders will be reimplemented in Metal
+3. Off screen targets will be implemented in Metal and exposed in a such a way they can be read from and shared
+4. Rendering passes will be implemented in Metal
+5. Snapshots will not block the rendering pipeline
+6. Atlases will be implemented for Metal and be accessible from any thread
```

Comment:
Constructing texture atlases on the GPU would be interesting!  And I guess you do part of that with a GPU based memory copy.  But with MapLibre Native's current implementation I think it will be hard to move too much more of it on the GPU.  And my sense is it's not necessary for performance.  Interesting idea though...

What I meant here is that you should be able to construct an Atlas on any thread without it having to do the copies on the main thread.